File: backend/src/main/java/com/etour/controller/ReviewController.java
package com.etour.controller;

import java.util.List;

import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.etour.entity.Review;
import com.etour.service.ReviewService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/reviews")
public class ReviewController {

private final ReviewService reviewService;
	
	public ReviewController(ReviewService reviewService) {
		this.reviewService = reviewService;
	}
	
	@GetMapping
	public List<Review> getAllReview(){
		return reviewService.getAllReview();
	}
	
	@GetMapping("{id}")
	public Review getById(@Valid @PathVariable("id") long id) {
		return reviewService.getById(id);
	}
	
	@PostMapping
	public Review addReview(@Valid @RequestBody Review review) {
		return reviewService.addReview(review);
	}
	
	@PutMapping("{id}")
	public Review editReview(@Valid @PathVariable("id") long id, @RequestBody Review review) {
		return reviewService.editReview(id, review);
	}
	
	@DeleteMapping("{id}")
	public void deleteReview(@Valid @PathVariable("id") long id) {
		reviewService.deleteReview(id);
	}
	
	@GetMapping("/customer/{id}")
	public List<Review> getByCustomerId(@Valid @PathVariable("id") long id){
		return reviewService.getByCustomerId(id);
	}
	
	@PostMapping("/customer/{customer_id}/tour/{tour_id}")
	public Review addByCustomerIdTourId(@Valid 
			@PathVariable("customer_id") long customer_id,
			@PathVariable("tour_id") long tour_id,
			@RequestBody Review review) {
		return reviewService.addByCustomerIdTourId(customer_id, tour_id, review);
	}
	
	@PutMapping("/customer/{customer_id}/tour/{tour_id}")
	public Review editByCustomerIdTourId(@Valid 
			@PathVariable("customer_id") long customer_id, 
			@PathVariable("tour_id") long tour_id, 
			@RequestBody Review review) {
		return reviewService.editByCustomerIdTourId(customer_id, tour_id, review);
	}
	
	@GetMapping("/tour/{id}")
	public List<Review> getByTourId(@Valid @PathVariable("id") long id){
		return reviewService.getByTourID(id);
	}
	
	@GetMapping("/rating/{id}")
	public List<Review> getByRating(@Valid @PathVariable("id") Integer id){
		return reviewService.getByRating(id);
	}
	
	@GetMapping("/rating/sort")
	public List<Review> sortByRating(){
		return reviewService.sortByRating();
	}
	
	@GetMapping("/tour/{id}/top5")
	public List<Review> top5Reviews(@PathVariable("id") long id){
		return reviewService.top5Reviews(id);
	}
	
	
	@GetMapping("/tour/{id}/avg")
	public Double getAverageRating(@PathVariable("id") long id) {
		return reviewService.getAverageRating(id);
	}
	
	
}

File: dotnet-backend/src/eTour.Api/Controllers/ReviewController.cs
using eTour.Application.Dtos;
using eTour.Application.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace eTour.Api.Controllers;

[ApiController]
[Route("api/reviews")]
public class ReviewController : ControllerBase
{
    private readonly IReviewService _service;

    public ReviewController(IReviewService service)
    {
        _service = service;
    }

    [HttpGet]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> GetAll(CancellationToken ct) => Ok(await _service.GetAllAsync(ct));

    [HttpGet("{id:long}")]
    [AllowAnonymous]
    public async Task<ActionResult<ReviewResponse>> GetById(long id, CancellationToken ct) => Ok(await _service.GetByIdAsync(id, ct));

    // ----- Moderation endpoints (ADMIN only) - act on any review by id. -----
    [HttpPost]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<ReviewResponse>> Add([FromBody] ReviewResponse review, CancellationToken ct) =>
        StatusCode(StatusCodes.Status201Created, await _service.AddReviewAsync(review, ct));

    [HttpPut("{id:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<ReviewResponse>> Edit(long id, [FromBody] ReviewRequest request, CancellationToken ct) =>
        Ok(await _service.EditReviewAsync(id, request, ct));

    [HttpDelete("{id:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> Delete(long id, CancellationToken ct)
    {
        await _service.DeleteReviewAsync(id, ct);
        return NoContent();
    }

    [HttpGet("customer/{id:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> GetByCustomerId(long id, CancellationToken ct) =>
        Ok(await _service.GetByCustomerIdAsync(id, ct));

    [HttpGet("tour/{id:long}")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> GetByTourId(long id, CancellationToken ct) =>
        Ok(await _service.GetByTourIdAsync(id, ct));

    [HttpGet("rating/{id:int}")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> GetByRating(int id, CancellationToken ct) =>
        Ok(await _service.GetByRatingAsync(id, ct));

    [HttpGet("rating/sort")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> SortByRating(CancellationToken ct) => Ok(await _service.SortByRatingAsync(ct));

    [HttpGet("tour/{id:long}/top5")]
    [AllowAnonymous]
    public async Task<ActionResult<IReadOnlyList<ReviewResponse>>> Top5Reviews(long id, CancellationToken ct) =>
        Ok(await _service.Top5ReviewsAsync(id, ct));

    [HttpGet("tour/{id:long}/avg")]
    [AllowAnonymous]
    public async Task<ActionResult<double>> GetAverageRating(long id, CancellationToken ct) => Ok(await _service.GetAverageRatingAsync(id, ct));

    // ----- Own-review endpoints - identity always from the JWT principal. -----
    [HttpPost("me/tour/{tourId:long}")]
    [Authorize(Roles = "CUSTOMER")]
    public async Task<ActionResult<ReviewResponse>> AddMyReview(long tourId, [FromBody] ReviewRequest request, CancellationToken ct) =>
        Ok(await _service.AddMyReviewAsync(tourId, request, ct));

    [HttpPut("me/tour/{tourId:long}")]
    [Authorize(Roles = "CUSTOMER")]
    public async Task<ActionResult<ReviewResponse>> EditMyReview(long tourId, [FromBody] ReviewRequest request, CancellationToken ct) =>
        Ok(await _service.EditMyReviewAsync(tourId, request, ct));

    [HttpDelete("me/{reviewId:long}")]
    [Authorize(Roles = "ADMIN,CUSTOMER")]
    public async Task<IActionResult> DeleteMyReview(long reviewId, CancellationToken ct)
    {
        await _service.DeleteMyReviewAsync(reviewId, ct);
        return NoContent();
    }
}
